doc/source/theory/fluid_dynamics/linear_solvers.rst

```diff
@@ -1,4 +1,154 @@
+
+Then, in the preconditioned Krylov solver, :math:`\mathcal{M}^{-1} y` is computed by calculating :math:`z = L^{-1} y` and then :math:`U^{-1} z`. Let us also define the following matrix:
+
+.. math::
+
+    \mathcal{R} = \mathcal{A} - L U
+
+The different versions of ILU arise with the different treatment of the non-zero entries of the matrix :math:`\mathcal{R}`:
+
+* The first ILU version is known as the *no-fill* or *ILU(0)* method. It is defined as any pair of matrices :math:`L` and :math:`U` so that the elements of :math:`\mathcal{R}` are zero in the same locations as the system matrix :math:`\mathcal{A}`. This version may be insufficient to yield an adequate rate of convergence for certain problems. In practice this leads to an algorithm where the pattern of :math:`LU` is equal to the zero pattern of :math:`\mathcal{A}`.
+
+* Another version is the *ILU(k)* which allows for *fill-in elements* and improves accuracy.  In fact, it consists of zeroing out all the fills of level :math:`k+1` or higher. In other words, it constructs :math:`LU` by setting the pattern of this matrix to be the pattern obtained in the decomposition of *ILU(k-1)*. The drawbacks of this approach are that the amount of fill-in and computational work are not predictable for :math:`k > 0`. Also, when the fill is large, this is closer to the traditional LU decomposition, which is expensive. In addition, one must keep in mind that if there are elements that are being dropped and have a significant value, this will lead to an inaccurate factorization and to a larger number of iterations.  An important concept when dealing with these methods is the *level of fill*, which is used when calculating the entries of the preconditioner. It is basically a way of identifying which elements should be preserved or dropped in the final preconditioner. All the fill-in elements whose level does not exceed :math:`k` are kept. 
+
+* The previous versions can be further improved by using a threshold that controls the fill-in according to the size of the elements and not only based on the sparsity pattern. One version is known as ILU with dual thresholding (*ILUT*).
+
+* Other versions try to reduce the effect of dropping some entries by taking into account the values in other ways. For example, one strategy is to add all the elements that have been dropped and subtract them from the diagonal entry of :math:`U`. This is another version called Modified ILU (MILU) factorization.
+
+* In Lethe we use the Trilinos implementation of the ILU method through `deal.II <https://www.dealii.org/>`_: `TrilinosWrappers::PreconditionILU <https://dealii.org/developer/doxygen/deal.II/classTrilinosWrappers_1_1PreconditionILU.html>`_. This is of type *ILU(k)* (where ILU(0) is also included). To use this preconditioner, one must specify the absolute and relative tolerance, and the fill level. If the iterative method fails, the fill level is increased by 1 and the solution is attempted again.
```

Comment:
I would explain what the absolute and relative tolerances are. These two parameters can have a very big impact on the convergence of the linear solver.